File: C/TCP-examples/tcp-5-server.cc
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <arpa/inet.h>
#include <sys/errno.h>
#include <string.h>
#include "support.h"

//
// Sample TCP server. This example does not fork a child for concurrency.
// Instead, it uses the "select" system call to wait for input
// on one of many possible sockets, including the listen socket.
// This allows a single threaded server to handle multiple clients concurrently.
//

const char *me = "SERVER";

int
main(int argc, char **argv)
{
  struct sockaddr_in addr;
  int listenfd;

  if ( argc != 3 ) {
    fprintf(stderr,"Usage: %s network port\n",
	    argv[0]);
  }

  fprintf(stdout, "%s: Specified network is *%s*\n", me, argv[1]);
  fprintf(stdout, "%s: Specified port is *%s*\n", me, argv[2]);


  listenfd = socket(AF_INET, SOCK_STREAM, 0);
  check_and_exit(listenfd, "socket");
  //
  // Initialize socket
  //
  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  inet_pton(AF_INET, argv[1], &addr.sin_addr);
  addr.sin_port = htons(atoi(argv[2]));

  //
  // Bind socket to specified address
  //
  int ret = bind(listenfd, (struct sockaddr *)&addr, sizeof(addr));
  check_and_exit(ret, "bind");

  //
  // Place socket in LISTEN state
  //
  ret = listen(listenfd, 100);
  check_and_exit(ret,"listen");

  //
  // Start the acceptance loop
  //

  int maxfd = -1;
  fd_set readfd;
  FD_ZERO(&readfd);
  FD_SET(listenfd, &readfd);
  maxfd = listenfd; 

  for(;;) {
    //
    fd_set tempfd;

    fd_copy(&tempfd, &readfd, maxfd+1);

    int n = select(maxfd + 1, &tempfd,
		   NULL, NULL, NULL);

    fprintf(stderr,"Got %d from select\n", n);

    for( int i = 0; i <= maxfd && n > 0; i++) {
      if ( FD_ISSET(i, &tempfd) ) {
	fprintf(stderr,"%s: File descriptor %d is ready\n", me, i);
	n--;
	if ( i == listenfd ) {
	  struct sockaddr_in theiraddr;
	  socklen_t theiraddr_len = sizeof(theiraddr);
	  int acceptfd = accept(listenfd, 
				(struct sockaddr *)&theiraddr,
				&theiraddr_len);
	  check_and_exit(acceptfd, "accept");
    
	  FD_SET(acceptfd, &readfd);
	  if ( acceptfd > maxfd ) {
	    maxfd = acceptfd;
	  }

	  fprintf(stderr,"%s: Start connection with...\n", me);
	  print_socket_endpoints(stderr,"Start connection with ..", acceptfd);
	} else {
	  int myfd = i;
	  char line[1024];
	  ssize_t r = read(myfd, line, sizeof(line));
	  if ( r < 0 ) {
	    perror("read");
	    FD_CLR(myfd, &readfd);
	    close(myfd);
	    break;
	  } else  if ( r ==  0 ) {
	    fprintf(stderr,"%s: ** End of input, exit\n", me);
	    FD_CLR(myfd, &readfd);
	    close(myfd);
	    break;
	  } else {
	    fprintf(stderr,"%s: ** %d Got %d bytes: %s\n", me, getpid(), (int) r, line);
	  }

	  int w = write(myfd, line, r);
	  if ( w < 0 ) {
	    perror("write");
	    break;
	  } else {
	    fprintf(stderr,"%s: ** %d sent %d bytes to socket\n", me, getpid(), w);
	  }

	}
      }
    }
  }
  fprintf(stderr,"%s: Parent is exiting...\n", me);
}

File: Python/TCP-examples/tcp-5-server.py
#!/usr/bin/env python3
##
## Corresponds to tcp-4-server.c. I omit tcp-3-server.py.
## Pair this with tcp-2-client
##
import socket
import sys,os,select


##
## Sample TCP server. This example does not fork a child for concurrency.
## Instead, it uses the "select" system call to wait for input
## on one of many possible sockets, including the listen socket.
## This allows a single threaded server to handle multiple clients concurrently
## in a single process with the shared state that implies.
##

if len(sys.argv) < 3:
    print("usage:", sys.argv[0], "<network> <port>")
    sys.exit(1)

network = sys.argv[1]
port = sys.argv[2]
me = "server"

##
## Print socket information
##
def print_socket_endpoints(name, sockfd):
    print("%s: Socket %s is connecting %s(me) and %s(them)" % (
        me, name, sockfd.getsockname(), sockfd.getpeername() ))

##
## Create a socket, bind to any interface, specific port
##
listenfd = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sendAddress = (network, int(port))
listenfd.bind(sendAddress)
listenfd.listen(100)

##
## Accept connections forever
##
readList = [ listenfd ]
while True:
    print(f"{me}: wait for select")
    ##
    ## Select takes 3 lists of sockets - the read, write & other lists.
    ## The process will pause until it's possible to read, write or
    ## receive other notifications (e.g. a closed socket) on those
    ## sockets.
    ##
    ## In the example below, we are simply waiting for any client socket
    ## to have available data.
    ##
    rfd, wfd, ofd = select.select(readList, [], [])

    for fd in rfd:
        if fd == listenfd:
            ## Listen to a new connection?
            print(f"{me}: Listen for connection")
            acceptfd, theirAddr = listenfd.accept()
            print_socket_endpoints("listen", acceptfd)            
            readList.append( acceptfd )
        else:
            ## process a client command
            line = fd.recv(8192)
            if len(line) == 0:
                print("End of input, exit")
                fd.close()
                readList.remove( fd )
            else:
                print(f"{me}: received {line.decode('utf-8')}")
                fd.send(line)
